sdks/python/apache_beam/pipeline.py

```diff
@@ -191,47 +189,73 @@ class TransformUpdater(PipelineVisitor): # pylint: disable=used-before-assignmen
+          # # TODO:  Support replacing PTransforms with multiple inputs.
           if len(inputs) > 1:
             raise NotImplementedError(
                 'PTransform overriding is only supported for PTransforms that '
                 'have a single input. Tried to replace input of '
                 'AppliedPTransform %r that has %d inputs',
-                transform_node, len(inputs))
-          transform_node.transform = replacement_transform
-          self.pipeline.transforms_stack.append(transform_node)
+                original_transform_node, len(inputs))
+
+          # We have to add the new AppliedTransform to the stack before expand()
+          # and pop it out later to make sure that parts get added correctly.
+          self.pipeline.transforms_stack.append(replacement_trasform_node)
 
           # Keeping the same label for the replaced node but recursively
-          # removing labels of child transforms since they will be replaced
-          # during the expand below.
-          self.pipeline._remove_labels_recursively(transform_node)
+          # removing labels of child transforms of original transform since they
+          # will be replaced during the expand below.
+          # This is needed in-case replacement contains children that have
+          # lables that conflicts with lables of the children of the original.
+          self.pipeline._remove_labels_recursively(original_transform_node)
 
           new_output = replacement_transform.expand(inputs[0])
-          if new_output.producer is None:
-            # When current transform is a primitive, we set the producer here.
-            new_output.producer = transform_node
+          replacement_trasform_node.add_output(new_output)
 
           # We only support replacing transforms with a single output with
           # another transform that produces a single output.
           # TODO: Support replacing PTransforms with multiple outputs.
-          if (len(transform_node.outputs) > 1 or
-              not isinstance(transform_node.outputs[None], PCollection) or
+          if (len(original_transform_node.outputs) > 1 or
+              not isinstance(
+                  original_transform_node.outputs[None], PCollection) or
               not isinstance(new_output, PCollection)):
             raise NotImplementedError(
                 'PTransform overriding is only supported for PTransforms that '
                 'have a single output. Tried to replace output of '
                 'AppliedPTransform %r with %r.'
-                , transform_node, new_output)
+                , original_transform_node, new_output)
 
           # Recording updated outputs. This cannot be done in the same visitor
           # since if we dynamically update output type here, we'll run into
           # errors when visiting child nodes.
-          output_map[transform_node.outputs[None]] = new_output
+          output_map[original_transform_node.outputs[None]] = new_output
```

Comment:
<!--thread_id:cc_164903025_t; commit:eb3d1308494b34cb40336a41f40937e9b74c266a; resolved:1-->
<!--section:context-quote-->
> **charlesccychen** wrote:
> Is there a test case we can add that failed before but passes with this change?

<!--section:body-->
This particular update (using a new node instead of updating the existing node) was not due to a breakage but to make the overriding more cleaner. I don't have a test case now but if there's a bug we have a better change of exposing that now.